### PR TITLE
Extract copy/free arrayCritical code from StandardAccessBarrier

### PR DESCRIPTION
**Pull request by LinHu2016**
-Extract copyArrayCritical copyBackArrayCritical, copyStringCritical
  and freeStringCritical code from StandardAccessBarrier to
  ObjectAccessBarrier(baseclass).

### CODE REVIEW

**Comment by LinHu2016** [comment]
@amicic Could you please review the changes? Thanks

File: runtime/gc_base/ObjectAccessBarrier.cpp [hunk 1]
```diff
@@ -109,6 +109,95 @@ MM_ObjectAccessBarrier::tearDown(MM_EnvironmentBase *env)
 {
 }
 
+void
+MM_ObjectAccessBarrier::copyArrayCritical(J9VMThread *vmThread, GC_ArrayObjectModel *indexableObjectModel,
+	J9InternalVMFunctions *functions, void **data, J9IndexableObject *arrayObject, jboolean *isCopy)
```
Comment (amicic) [inline runtime/gc_base/ObjectAccessBarrier.cpp:114]: minor, i'd probably prefer not passing indexableObjectModel, but rather fetch it from _ext again
@dmitripivkine, opinion?

File: runtime/gc_base/ObjectAccessBarrier.cpp [hunk 2]
```diff
@@ -109,6 +109,95 @@ MM_ObjectAccessBarrier::tearDown(MM_EnvironmentBase *env)
...
+	if (vmThread->jniCriticalCopyCount > 0) {
+		vmThread->jniCriticalCopyCount -= 1;
+	} else {
+		Assert_MM_invalidJNICall();
+	}
+}
+
+void
+MM_ObjectAccessBarrier::copyStringCritical(J9VMThread *vmThread, GC_ArrayObjectModel *indexableObjectModel,
+	J9InternalVMFunctions *functions, jchar **data, J9JavaVM *javaVM,
```
Comment (amicic) [inline runtime/gc_base/ObjectAccessBarrier.cpp:157]: like with indexableObjectModel, probably cleaner not to pass javaVM

File: runtime/gc_base/ObjectAccessBarrier.cpp (hunk 1 above)
Comment (dmitripivkine) [reply runtime/gc_base/ObjectAccessBarrier.cpp:114]: I agree, let's get it from _ext

File: runtime/gc_modron_standard/StandardAccessBarrier.cpp [hunk 3]
```diff
@@ -408,43 +397,27 @@ MM_StandardAccessBarrier::jniReleasePrimitiveArrayCritical(J9VMThread* vmThread,
 {
 	J9JavaVM *javaVM = vmThread->javaVM;
 	J9InternalVMFunctions *functions = javaVM->internalVMFunctions;
```
Comment (amicic) [inline runtime/gc_modron_standard/StandardAccessBarrier.cpp:399]: I feel like this is not used any more here (but only in new helper methods) or any other get/release method. If so, we should not have it here and then not to pass it, but let the helper fetch it.

File: runtime/gc_base/ObjectAccessBarrier.cpp [hunk 4]
```diff
@@ -109,6 +109,95 @@ MM_ObjectAccessBarrier::tearDown(MM_EnvironmentBase *env)
...
+	J9IndexableObject *valueObject, J9Object *stringObject, jboolean *isCopy, bool isCompressed)
```
[the remark this answers is not in the document]
Comment (amicic) [reply runtime/gc_base/ObjectAccessBarrier.cpp:158]: Standard does not, but seems like VLHGC will

**Comment by amicic** [comment]
jenkins test sanity xLinux,aix jdk11

File: runtime/gc_base/ObjectAccessBarrier.hpp [hunk 5]
```diff
@@ -84,6 +84,49 @@ class MM_ObjectAccessBarrier : public MM_BaseVirtual
...
+	 * Copy string bytes to the allocated native memory.
+	 *
+	 * @param vmThread[in] the current J9VMThread.
+	 * @param data[in] the pointer of native memory address.
+	 * @param valueObject[in] the pointer of array object
+	 * @param stringObject[in] Java String
+	 * @param isCopy[out] if it true, the array is successfully copied.
+	 * @param isCompressed[in] if it is compressed.
+	 */
+	void copyStringCritical(J9VMThread *vmThread, jchar **data
```
Comment (amicic) [inline runtime/gc_base/ObjectAccessBarrier.hpp:118]: missing comma after data

File: runtime/gc_base/ObjectAccessBarrier.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (amicic) [reply runtime/gc_base/ObjectAccessBarrier.cpp:159]: so, i'd just leave it for now, and when we review VLHGC, we'll decide if we should pass or not

**Comment by amicic** [comment]
jenkins test sanity xLinux,aix jdk11

**Comment by amicic** [comment]
@keithc-ca @pshipton

sanity.functional fails 
```
22:21:14      [javac] /home/jenkins/workspace/Test_openjdk11_j9_sanity.functional_ppc64_aix_Personal_testList_1/aqa-tests/functional/JLM_Tests/src/org/openj9/test/java/lang/management/TestOpenJ9DiagnosticsMXBean.java:510: error: cannot find symbol
22:21:14      [javac] 		String newDumpOptions = diagBean.getDumpOptions();
```

I see it recently changed
 https://github.com/eclipse-openj9/openj9/pull/20604

**Comment by amicic** [comment]
Internal Axxon build 82030 passed. Ignoring sanity.functional failure and merging...

**Review by amicic — APPROVED** [review]

**Comment by pshipton** [comment]
I think it's a timing thing. New test material with older JVM.

**Pull request merged**